rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,385 @@
+formats that Kubernetes itself supports.
+
+Generally, most Kubernetes setups will put the logs at the `kubelet`-configured
+locations in a .
+
+There is [official documentation][k8s_log_path_location_docs] at Kubernetes
+project regarding logging. I had a misconception that it specifies reading these
+log files as an explicitly supported way of consuming the logs, however, I
+couldn't find a confirmation of that when I checked.
+Nonetheless, Kubernetes log files is a de-facto well-settled interface, that we
+should be able to use reliably.
+
+#### File locations
+
+We can read container logs directly from the host filesystem. Kubernetes stores
+logs such that they're accessible from the following locations:
+
+- [`/var/log/pods`][var_log_pods_src];
+- `/var/log/containers` - legacy location, kept for backward compatibility
+  with pre `1.14` clusters.
+
+To make our lives easier, here's a [link][build_container_logs_directory_src] to
+the part of the k8s source that's responsible for building the path to the log
+file. If we encounter issues, this would be a good starting point to unwrap the
+k8s code.
+
+#### Log file format
+
+As already been mentioned above, log formats can vary, but there are certain
+invariants that are imposed on the container runtimes by the implementation of
+Kubernetes itself.
+
+A particularity interesting piece of code is the [`ReadLogs`][k8s_src_read_logs]
+function - it is responsible for reading container logs. We should carefully
+inspect it to gain knowledge on the edge cases. To achieve the best
+compatibility, we can base our log files consumption procedure on the logic
+implemented by that function.
+
+Based on the [`parseFuncs`][k8s_src_parse_funcs] (that
+[`ReadLogs`][k8s_src_read_logs] uses), it's evident that k8s supports the
+following formats:
+
+- Docker [JSON File logging driver] format - which is essentially a simple
+  [`JSONLines`][jsonlines] (aka `ndjson`) format;
+- [CRI format][cri_log_format].
+
+We have to support both formats.
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+1. [Fluentd k8s integration]
+1. [LogDNA k8s integration]
+1. [Honeycomb integration]
+1. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+1. [Influx Helm charts]
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community-driven
+   integrations.
+
+## Outstanding Questions
+
+### From Ben
+
+1. What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].
+1. What is the best to avoid Vector from ingesting it's own logs? I'm assuming
```

Comment:
If we want to ignore the logs at the source level, we only have access to:
- pod namespace
- pod name
- pod UID
- container name

This is quite limiting.

In the vector code, we probably want to be as flexible as possible, and support filtering by every combination of the above.

In our recommended YAML files, we can filter using namespace, since we deploy into a dedicated namespace there anyway. But it's important our users can switch to ignoring by the pod uids, or other way they please.

Reply:
Explained this in more detail at the RFC itself.